# RAGAS Before-and-After Comparison

This notebook compares the baseline Networking RAG system (Phase 2) and the improved Networking RAG system (Phase 4) using the same RAGAS evaluation metrics.

In [2]:
 import pandas as pd

 print(
        "Libraries imported successfully!"
 )


Libraries imported successfully!


In [3]:
# Load evaluation result files

try:

    phase2_df = pd.read_csv(
        "phase2_ragas_results.csv"
    )

    phase4_df = pd.read_csv(
        "phase4_ragas_results.csv"
    )

    print(
        "Files loaded successfully!"
    )

except Exception as e:

    print(
        "Loading Error:"
    )

    print(e)

Files loaded successfully!


In [4]:
# Verify loaded files

try:

    print(
        "\nPhase 2 Results:"
    )

    print(
        phase2_df.head()
    )

    print(
        "\nPhase 4 Results:"
    )

    print(
        phase4_df.head()
    )

except Exception as e:

    print(e)


Phase 2 Results:
                     user_input  \
0                  What is TCP?   
1                  What is UDP?   
2                  What is DNS?   
3  What is the purpose of HTTP?   
4        What is the OSI model?   

                                  retrieved_contexts  \
0  ['The Transmission Control Protocol (TCP) is o...   
1  ["In computer networking, the User Datagram Pr...   
2  ['The Domain Name System (DNS) is a hierarchic...   
3  ['HTTP (Hypertext Transfer Protocol) is an app...   
4  ['The Open Systems Interconnection (OSI) model...   

                                            response  \
0  TCP stands for Transmission Control Protocol, ...   
1  UDP stands for User Datagram Protocol, which i...   
2  The Domain Name System (DNS) is a hierarchical...   
3  The purpose of HTTP (Hypertext Transfer Protoc...   
4  The Open Systems Interconnection (OSI) model i...   

                                           reference  faithfulness  \
0  TCP is a reliable connec

In [5]:
# Calculate average metric scores

try:

    metrics = [

        "faithfulness",

        "answer_relevancy",

        "context_precision",

        "context_recall"

    ]

    phase2_scores = {}

    phase4_scores = {}

    for metric in metrics:

        phase2_scores[
            metric
        ] = phase2_df[
            metric
        ].mean()

        phase4_scores[
            metric
        ] = phase4_df[
            metric
        ].mean()

    print(
        "Average scores calculated successfully!"
    )

except Exception as e:

    print(e)

Average scores calculated successfully!


In [6]:
# Display average scores

try:

    print(
        "\nPhase 2 Scores\n"
    )

    print(
        phase2_scores
    )

    print(
        "\nPhase 4 Scores\n"
    )

    print(
        phase4_scores
    )

except Exception as e:

    print(e)


Phase 2 Scores

{'faithfulness': np.float64(0.9166666666666666), 'answer_relevancy': np.float64(0.7907639009412731), 'context_precision': np.float64(0.9999999999713334), 'context_recall': np.float64(1.0)}

Phase 4 Scores

{'faithfulness': np.float64(1.0), 'answer_relevancy': np.float64(0.803989889524934), 'context_precision': np.float64(0.9258333333131874), 'context_recall': np.float64(1.0)}


In [7]:
# Create before-and-after comparison table

try:

    comparison_df = pd.DataFrame({

        "Metric": [

            "Faithfulness",

            "Answer Relevancy",

            "Context Precision",

            "Context Recall"

        ],

        "Phase 2": [

            phase2_scores[
                "faithfulness"
            ],

            phase2_scores[
                "answer_relevancy"
            ],

            phase2_scores[
                "context_precision"
            ],

            phase2_scores[
                "context_recall"
            ]

        ],

        "Phase 4": [

            phase4_scores[
                "faithfulness"
            ],

            phase4_scores[
                "answer_relevancy"
            ],

            phase4_scores[
                "context_precision"
            ],

            phase4_scores[
                "context_recall"
            ]

        ]

    })

    comparison_df[
        "Improvement"
    ] = (

        comparison_df[
            "Phase 4"
        ]

        -

        comparison_df[
            "Phase 2"
        ]

    )

    print(
        comparison_df
    )

except Exception as e:

    print(e)

              Metric   Phase 2   Phase 4  Improvement
0       Faithfulness  0.916667  1.000000     0.083333
1   Answer Relevancy  0.790764  0.803990     0.013226
2  Context Precision  1.000000  0.925833    -0.074167
3     Context Recall  1.000000  1.000000     0.000000


In [8]:
# Create before-and-after comparison table

try:

    comparison_df = pd.DataFrame({

        "Metric": [

            "Faithfulness",

            "Answer Relevancy",

            "Context Precision",

            "Context Recall"

        ],

        "Phase 2": [

            phase2_scores[
                "faithfulness"
            ],

            phase2_scores[
                "answer_relevancy"
            ],

            phase2_scores[
                "context_precision"
            ],

            phase2_scores[
                "context_recall"
            ]

        ],

        "Phase 4": [

            phase4_scores[
                "faithfulness"
            ],

            phase4_scores[
                "answer_relevancy"
            ],

            phase4_scores[
                "context_precision"
            ],

            phase4_scores[
                "context_recall"
            ]

        ]

    })

    comparison_df[
        "Improvement"
    ] = (

        comparison_df[
            "Phase 4"
        ]

        -

        comparison_df[
            "Phase 2"
        ]

    )

    print(
        comparison_df
    )

except Exception as e:

    print(e)

              Metric   Phase 2   Phase 4  Improvement
0       Faithfulness  0.916667  1.000000     0.083333
1   Answer Relevancy  0.790764  0.803990     0.013226
2  Context Precision  1.000000  0.925833    -0.074167
3     Context Recall  1.000000  1.000000     0.000000


**Evaluation Summary**

The baseline (**Phase 2**) and improved (**Phase 4**) RAG systems were evaluated using **RAGAS metrics** on the same set of evaluation questions. A comparison table was created to measure the impact of **Better Chunking** and **Hybrid Search**.

**RAGAS Metrics**

• **Faithfulness** – Checks whether the generated answer is supported by the retrieved context.

• **Answer Relevancy** – Measures how well the answer addresses the user's question.

• **Context Precision** – Measures the relevance of the retrieved chunks.

• **Context Recall** – Measures whether all relevant information was retrieved.


**Results**

• **Faithfulness:** 0.9167 → 1.0000

• **Answer Relevancy:** 0.7908 → 0.8040

• **Context Precision:** 1.0000 → 0.9258

• **Context Recall:** 1.0000 → 1.0000

The improved system achieved better answer grounding and maintained perfect context recall. The slight decrease in **Context Precision** is expected because **Hybrid Search** retrieves a broader set of relevant chunks by combining vector search and BM25 retrieval, improving coverage while introducing a small amount of additional context.
